Add justification controls to constrained layout

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Fixes #43650.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Adds justification controls to the "constrained" layout type. These change the `margin` values on the block children so block content can be aligned left or right.

Note that setting justification to "right" does not right-align any text content, only the child blocks themselves.

Note also that when setting "left" or "right" justification on a Post Content block, the post editor view won't reflect that layout because we don't yet have a way to access the Post Content block settings from the post editor.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. In the Post Content block sidebar Layout section, toggle on "Inner blocks use content width";
2. Check that "Justification" controls are added to the bottom of the Layout section;
3. Change justification value and check that new value is reflected in the editor and on the front end.

## Screenshots or screencast <!-- if applicable -->

Before (content centred):
<img width="1121" alt="Screen Shot 2022-09-12 at 1 51 16 pm" src="https://user-images.githubusercontent.com/8096000/189570942-4df107e5-cb6e-41f4-be8c-0a12ce1d07a4.png">

After (content left-aligned):
<img width="1108" alt=" wat" src="https://user-images.githubusercontent.com/8096000/189570995-5baed65c-41e1-412a-b8b3-54735b5526a6.png">

tellthemachines committed:
Add justification controls to constrained layout

carolinan commented:
This mostly tests well for me.

When the post content block is placed inside a group with "Inner blocks use content width" enabled, the justification options for the post content block doesn't seem to do anything.

I believe the UI may need some small changes, because it may look like the "Customize the width for all...." text is related to the justification option:
<img width="325" alt="The help text of the first option sits close to the label for the justification option " src="https://user-images.githubusercontent.com/7422055/189593445-1e60033f-ac4f-45c8-b004-6ff8a71c4741.png">

tellthemachines commented:
Thanks for testing @carolinan !

> When the post content block is placed inside a group with "Inner blocks use content width" enabled, the justification options for the post content block doesn't seem to do anything.

Yeah, this will always happen if we nest two blocks with the same content width. Not sure what would be best to do - maybe show a warning that the parent is already constrained? That would require adding some sort of context mechanism so we could know the parent's layout settings.

> I believe the UI may need some small changes, because it may look like the "Customize the width for all...." text is related to the justification option:

Yes, I just threw it in there, it should probably get some design feedback 😅 cc. @WordPress/gutenberg-design

jasmussen commented:
Thanks for working on this! Here's a GIF:

![status](https://user-images.githubusercontent.com/1204802/189602992-0a2674a7-a9be-4f02-808f-f8ef02274b06.gif)

Here's a "before" shot for reference:

<img width="1392" alt="before" src="https://user-images.githubusercontent.com/1204802/189603361-c3c1d4ff-6df4-4727-95a3-edaba204bb96.png">

This is an important feature, thank you for working on it (and thanks for pinging the design group!)

The big challenge to address here is that the Layout panel is already both big, and complex, so we need to simplify and reduce if possible. Here's a few minimal things we can do:

<img width="688" alt="Layout Panel v1" src="https://user-images.githubusercontent.com/1204802/189606074-a01c5fd6-fd0a-43bc-8679-6b356c387a15.png">

* Remove the "reset" button. It does the same as clearing out both values.
* Use the right gray help-text style. The toggle has a prop that outputs the correct styling, I'm not sure what the best way is for the two width controls, but we should avoid just applying the CSS class (CC: @ciampo @mirka for input)
* Use the correct label style and apperance for the justification controls. It looks like this uses a fieldset/legend combo, unlike for example "Letter case". This may be a components issue too. Should the legend element just use the same allcaps style as labels?
* The spacing is a bit off compared to [the grid](https://github.com/WordPress/gutenberg/issues/43423#issuecomment-1232540038). Should be 16px between sections of a panel, 8px between label/input/help text.

Letter case, for reference:
<img width="304" alt="Screenshot 2022-09-12 at 10 18 37" src="https://user-images.githubusercontent.com/1204802/189606951-360fa282-3253-4911-899d-61267dcc7a5a.png">

A followup could be to move this panel to use the ToolsPanel component. This would allow us to hide the justification control by default. The control is useful when you need it, but probably not always?

<img width="688" alt="Layout Panel v2" src="https://user-images.githubusercontent.com/1204802/189607236-127cbee3-9fa0-4d6b-b997-d79ad337009b.png">

ciampo commented:
> Use the right gray help-text style. The toggle has a prop that outputs the correct styling, I'm not sure what the best way is for the two width controls, but we should avoid just applying the CSS class (CC: @ciampo @mirka for input)

This is a tricky one, since in this in case we'd like to apply one help text paragraph to two separate controls ("content" and "wide"), and I don't think that's possible with out of the box components. It _may_ be possible to achieve something similar with [`BaseControl`](https://wordpress.github.io/gutenberg/?path=/story/components-basecontrol--with-help-text), but I'm not sure how we could represent these two separate controls correctly in terms of semantics.

Not sure if it's a good idea, but we could consider creating some sort of "help text" component ? @mirka do you have any ideas?

jasmussen commented:
Just for the purposes of keeping things simple, if we can just get the right color and spacing, that would go a long way.

mtias commented:
This is great, thanks for implementing it quickly.

jameskoster commented:
> When the post content block is placed inside a group with "Inner blocks use content width" enabled, the justification options for the post content block doesn't seem to do anything.

This just highlights how awkward the current layout controls are, and how bad things can get when you start nesting containers. A similar issue exists on trunk when child containers set different content-width values to their parents. On balance it is probably an issue to explore separately?

In terms of the work in this PR, Joen's suggestions look good. 

I wonder if it would be possible to reflect the justification in the post editor? Here's my current setup:

| Single template | Front end | Post Editor |
| --- | --- | --- |
| <img width="1576" alt="Screenshot 2022-09-12 at 12 28 27" src="https://user-images.githubusercontent.com/846565/189642518-b6bf7aad-6501-442e-8dd8-a6f124cbf135.png"> | <img width="1576" alt="Screenshot 2022-09-12 at 12 28 47" src="https://user-images.githubusercontent.com/846565/189642530-c4aeb36b-7450-4e20-a2fa-21a537a5078f.png"> | <img width="1574" alt="Screenshot 2022-09-12 at 12 29 06" src="https://user-images.githubusercontent.com/846565/189642542-61fad831-f5fa-49ad-949b-8631448aece4.png"> |

The Post Editor is letting the team down by still justifying everything centrally :)

mirka commented:
@tellthemachines What do you think about implementing this using ToggleGroupControl? We recently [added support](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-togglegroupcontrol--borderless) for this exact use case (i.e. icon buttons, borderless, `radio`-like behavior).

mirka commented:
> Not sure if it's a good idea, but we could consider creating some sort of "help text" component ?

@ciampo It's probably a good idea! We can expose it as a subcomponent of BaseControl, similar to `BaseControl.VisualLabel` The main use case would be to do something like this, which I think is applicable to this particular case as well:

```jsx
<fieldset>
  <SomeControl />
  <AnotherControl />
  <BaseControl.VisualHelpText as="legend">
    The help text.
  </BaseControl.VisualHelpText>
</fieldset>
```

andrewserong commented:
This is testing nicely for me, too! The only other tiny issue I ran into that I didn't see mentioned above is being able to toggle _off_ the justification (e.g. clicking the center button explicitly sets `"justifyContent":"center"` in the block comment in the markup, rather than being able to remove it altogether). However, I don't think this is a blocker as we don't do anything with this in the server-rendered output, and toggling off "Inner blocks use content width" effectively resets this value anyway.

> I wonder if it would be possible to reflect the justification in the post editor?

I think this would be great to investigate separately — if we think of this as effectively a bug in the post editor styling then perhaps we can argue that we can improve this beyond the feature freeze? One potential way to explore addressing this might to be see if we can extract the `post-content` block used in the selected template for the post currently being edited. If we can gather the attributes from that block instance and apply them to the post editor styling, then that might get us closer to a 1:1 between the editor and site frontend. Not sure how do-able that is, but I'm happy to write up an issue for it if there isn't one already 🙂

tellthemachines committed:
Improve layout controls panel

tellthemachines commented:
Thanks for the feedback folks! 

@jasmussen I think I've addressed all your comments except for the spacing between the toggle and the inputs, because it's coming from base control styles:

<img width="672" alt="Screen Shot 2022-09-13 at 11 43 24 am" src="https://user-images.githubusercontent.com/8096000/189801098-e10dfc43-f6c8-4c0e-968e-0313ed8ba148.png">

Changing that value will probably impact other areas of the UI, not sure if we want that? Alternatively I can add a local override.

@mirka I was able to use `ToggleGroupControl`, thanks for the suggestion!

The remaining problem with the help text for the input fields is that it's not semantically connected to them in any way, so essentially useless for a screen reader user. I could wrap them in a `fieldset` and make the text a `legend`, but we'll need to either switch the text to above the controls or deploy some very awkward positioning tricks so that `legend` can be the first child of `fieldset`. Not sure if there's any other way to label a group of fields - from the comments I assume there's nothing we can use in the components package?

>> I wonder if it would be possible to reflect the justification in the post editor?

> I think this would be great to investigate separately — if we think of this as effectively a bug in the post editor styling then perhaps we can argue that we can improve this beyond the feature freeze? One potential way to explore addressing this might to be see if we can extract the post-content block used in the selected template for the post currently being edited. If we can gather the attributes from that block instance and apply them to the post editor styling, then that might get us closer to a 1:1 between the editor and site frontend. Not sure how do-able that is, but I'm happy to write up an issue for it if there isn't one already 🙂

I created #42911 a while ago; it's specifically about knowing the layout type of Post Content, but we can use it to track the whole thing. Comments and ideas welcome!

tellthemachines committed:
Override base component bottom margin

andrewserong reviewed (approved):
Nice work on the quick follow-ups, @tellthemachines! The changes looks nice in the editor for me, and it's testing well:

![2022-09-13 14 02 44](https://user-images.githubusercontent.com/14988353/189805425-f0511786-9610-411c-8f89-fd58c67a1a2e.gif)

With the `ToggleGroupControl` it might be worth us also updating the Flex layout controls and labels in a separate PR for consistency in the UI? Definitely not a blocker for this PR.

This LGTM from my perspective and looks like it implements the design feedback so far 🎉 (looks much nicer to me without the reset button 👍). I agree with Joen, that it'd be good for us to also look at implementing the ToolsPanel for layout controls one day, particularly as we add more controls to the UI.

> I created https://github.com/WordPress/gutenberg/issues/42911 a while ago; it's specifically about knowing the layout type of Post Content, but we can use it to track the whole thing. Comments and ideas welcome!

Ah, perfect, I knew you'd written something up already. I'll add some thoughts / ideas to that issue 🙂

noisysocks commented:
Are you meant to be able to remove the justification? Unfortunately `ToggleGroupControl` doesn't support that right now 😀

tellthemachines commented:
> Are you meant to be able to remove the justification?

Good question! We're not doing it with the flex layout controls, which these are based on. I can't think of any real world use case for it, as the default state will always be one of the possible settings. It might even be _less_ confusing to make sure that it's always clear which setting is applied by default?

jameskoster commented:
The icons in the justification control appear to be the wrong color (they are a lighter grey than all other icon buttons). Is that something that occurred in this PR, or something relating to the underlying component? Either way it would be good to fix it because at the moment the buttons look disabled 🙈

jasmussen commented:
I think those icons are that shade of gray because it's a segmented control under the hood, and the segmented control has a #757575 color gray for non-active items. However I'd agree, when we use icons, we should probably just keep the color dark gray. Sorry about the trouble, @ciampo @mirka

paaljoachim commented:
Testing. Using the PR Gutenberg build.
It seems that this is for FSE themes.
Activating Twenty Twenty Two.
Going into the site editor and adding the Post Content block.

> In the Post Content block sidebar Layout section, toggle on "Inner blocks use content width";

Ok.

> Check that "Justification" controls are added to the bottom of the Layout section;

I can see these.

> Change justification value  and check that new value is reflected in the editor and on the front end.

**Changing to left.**

![Screenshot 2022-09-13 at 11 11 54](https://user-images.githubusercontent.com/5323259/189862826-78430ee6-cb8e-4d0f-b9b4-d239129a9ac5.jpg)

**Changing to center.**

![Screenshot 2022-09-13 at 11 12 03](https://user-images.githubusercontent.com/5323259/189862885-aeb78e5f-0b9d-4832-86e8-e71cff570d34.jpg)

**Changing to right.**

![Screenshot 2022-09-13 at 11 13 36](https://user-images.githubusercontent.com/5323259/189862933-3ebb1a0d-23a2-43e3-97e3-c4a30dfb36a1.jpg)


### Frontend.

**Left.**
![Screenshot 2022-09-13 at 11 18 16](https://user-images.githubusercontent.com/5323259/189863438-797d833d-b037-4ba9-aea2-df4ea8090d49.jpg)

**Center.**
![Screenshot 2022-09-13 at 11 18 30](https://user-images.githubusercontent.com/5323259/189863481-9ac163fb-9a42-497c-b2d1-eb21e3559118.jpg)

**Right.**
![Screenshot 2022-09-13 at 11 18 41](https://user-images.githubusercontent.com/5323259/189863518-e7665062-dc33-470c-be5a-8830d4ca25d5.jpg)

------------

### What I see...

There are hardly any differences in the backend between using left/center/right justification. I expected it to be related to the boundry it is in. I also expected even bigger differences on the frontend.

(For whatever reason when I left the site editor and came back in the Justification controls did not work....Skipping this part.)

Can we change the color of the grey controls to become black similar to the icons for Content and Wide?
As grey icons to me gives a signal that these are not in use/deactivated.

![Screenshot 2022-09-13 at 11 30 17](https://user-images.githubusercontent.com/5323259/189866156-91f062a7-9a03-4158-b338-4a0e41b4ed30.jpg)

tellthemachines commented:
> There are hardly any differences in the backend between using left/center/right justification. I expected it to be related to the boundry it is in. I also expected even bigger differences on the frontend.

@paaljoachim thanks for testing! If the editor canvas width is not much more than the content width, it might be hard to tell the difference between the three settings. On the front end it will also depend on viewport width. As I mentioned in the PR description, this control doesn't align text content, only the blocks themselves. So depending on the width of the content vs the viewport, plus any root padding, the difference might be quite subtle.
